### PR TITLE
Show getting started widget on dashboard (on localhost)

### PR DESCRIPTION
**Pull request by tabacitu**
## WHY

Addresses https://github.com/Laravel-Backpack/ideas/issues/170

### BEFORE - What was wrong? What was happening before this PR?

After you installed Backpack for the first time, you were greeted by a blank dashboard, that did nothing to guide you on your next steps:

CleanShot 2022-08-04 at 16 47 59@2x

### AFTER - What is happening after this PR?

On localhost, we show a different widget instead of the plain one, that instructs devs on the 3 important steps that would introduce them to Backpack:

CleanShot 2022-08-04 at 16 49 24@2x

### How did you achieve that, in technical terms?

Added a `view` widget instead of the current one, when needed.

### Is it a breaking change?

Yes and no... I mean... it's non-breaking, because it's only visual and only on localhost. But it's also breaking, because it's visual and it'll happen to existing projects that haven't customized their dashboard... To get rid of this, devs have to go to their config file and set it to `false`...

### Is this ready?

I don't know... Let's call it an advanced first draft. I think it goes a long way towards introducing people to Backpack but... it's not as good as it could be... then again, it could take us a few weeks to polish this and create videos for it. So until then... this first draft is better than nothing...

CleanShot 2022-08-04 at 16 56 52@2x

Then if you click the last link here, new text will appear, to make some customizations:

CleanShot 2022-08-04 at 16 57 48@2x

And the other two accordion items are pretty basic:
CleanShot 2022-08-04 at 16 58 19@2x

And

CleanShot 2022-08-04 at 16 58 33@2x

---

What I don't know right now is if... I didn't go a bit overboard with explaining how to do password/unique validation, and using Model Events... Not sure if this qualifies as crash course (good) or too much, too soon (bad). 

I'll sleep on it. Eager to hear what others think too.

### CODE REVIEW

**Comment by pxpm** [comment]
Hey @tabacitu 

I like the idea, I think it describes very well what should be the baby-steps to start interacting with Backpack. 
I am with you on the "maybe a little too much" for the UsersController. 
1- creating the controller ? ✅ 
2- Validation ? ✔️ | ❌ (If anything, I think we should ONLY instruct developers to use the FormRequests. It addds alot of functionality not directly available in fields like the `$stopOnFirstFailure`, to where should user be redirected, preparing the validation etc. And it's the default Laravel way! All of this topics ofcourse are out of scope for this get started).
3- Password management? ❌ It's too many concepts and not directly related with Backpack in a "getting started".

I get why you pick the User model, it's already there ... But it's quite a complicated one to deal because of all the business logic it involves. 

I would suggest we create something to add into users, like `Address` (hasOne) or `Pets` (hasMany). It would need one aditional step (create the migration), but we would show Backpack relationships (feature flag of this software).

We could then tell developers: now that you have this new Address for Users, it would be TOP if you could add the address directly when creating a new User, rigth ? We got you covered, there are two ways to proceed here:
a) we already did some of the hard work for you on the user management side, if you want to have user management in your Backpack application plus Roles and Permissions go ahead now and install `backpack permission manager package`.
b) If you simply want to have user management (without roles and permissions) follow me on how to create the User crud Controller: 
....
> I would tell everything as you did but would omit the password management part, make a note that it's required to properly work and point to permission manager crud controller where we already do that and keep that "simple" here. 

**Now lets add our Address relation into user controller**
If you followed:
a) tell them to create a new controller, explain them how to override a backpack package controller to customize stuff etc. (probably point to a FAQ we should have online with exactly this question).
b) head to user controller and add in `setupCreateOperation()` add `CRUD::field('address.street')` and in `ListOperation` `CRUD::column('address.street')`. 

Very important lessons teached here and keeping a similar approach as you did. 

Let me know what you think, if we can refine this idea.

Cheers

**Comment by tabacitu** [comment]
> I get why you pick the User model, it's already there ... But it's quite a complicated one to deal because of all the business logic it involves.

I know... that's the downside of it...

> I would suggest we create something to add into users, like Address (hasOne) or Pets (hasMany). It would need one aditional step (create the migration), but we would show Backpack relationships (feature flag of this software).

Creating a migration is a BIG extra step... that's not actually Backpack-specific. That's what annoys me. But you're right, Users don't show relationship fields... which IS in fact a big selling point... Damn...

> Let me know what you think, if we can refine this idea.

Ok so I've decided. I don't like creating an Addresses CRUD. Because 99% of all projects won't need `Address`. So they have no reason to create it - just for learning, then they delete it... I wouldn't do that myself. And I expect it will deter people from actually following the steps. They'll think "_Address... I don't need address... shut up_".

That's what I liked creating a CRUD for `User`. In addition to the fact that the Model is already there, developers will probably keep the Users CRUD afterwards. And it's easier to convince people to follow our steps... if they were going to build that CRUD anyway. If it'll be useful to them, since 90% of all projects don't actually need roles and permissions - a Users CRUD is enough.

> Now lets add our Address relation into user controller
> If you followed:
> a) tell them to create a new controller, explain them how to override a backpack package controller to customize stuff etc. (probably point to a FAQ we should have online with exactly this question).
> b) head to user controller and add in setupCreateOperation() add CRUD::field('address.street') and in ListOperation CRUD::column('address.street').

No way, Jose. That's less useful to learn and it doesn't concern the main package (CRUD). It concerns an addon. The first things we should teach are the first things they should know - how to generate stuff, how to use Operations, how to use fields/columns. Everything else is "advanced", not "basic".

---

No, I'm convinced. **We stay with Users - it's not perfect, but it's the best compromise.** We've already hidden the complex stuff, and we can do a video for those later, to replace all the fluff.

PROs:
- model is already there (fewer complex instructions)
- devs are more likely to try it
- it's actually useful (devs can keep it)

CONs:
- doesn't show relationship fields;
- to make it work well, it needs quite a bit of editing for the Create and Update operations;

I think the PROs heavily outweigh the CONs.

---

**My question for you here Pedro was less about the tutorial. And more about the implementation.** About when we show this widget - basically _always_ on localhost, unless the dev has changed the config. This is OK for new projects in my opinion... but has implications for old projects - they'll start seeing this too... Just on locahost, but... it'll be a bad surprise, right? Then again, the widget itself shows how to hide it soo... maybe it isn't that bad? **What do you think** @pxpm ?

File: src/resources/views/base/dashboard.blade.php [hunk 1]
```diff
@@ -1,14 +1,21 @@
...
+        ];
+    } else {
+        $widgets['before_content'][] = [
+            'type'        => 'jumbotron',
+            'heading'     => trans('backpack::base.welcome'),
+            'content'     => trans('backpack::base.use_sidebar'),
+            'button_link' => backpack_url('logout'),
+            'button_text' => trans('backpack::base.logout'),
+        ];
+    }
```
Comment (tabacitu) [inline src/resources/views/base/dashboard.blade.php:17]: @pxpm heads-up, I am merging this later today. Last chance to stop me from doing something potentially stupid here 😅

**Comment by pxpm** [comment]
Let's just see if there is something more we can do here. As it is I am a little scary to approve this 🙃 

We can try to make it work, but how we are implementing this, we are bound to break people applications in certain conditions. 

I see the following scenarios:
- new install ✅ 
- old install without changes in dashboard ✅ (if they didn't change nothing in dashboard, probably they just removed the dashboard url and/or are not using the backpack dashboard page)
- old install and changed `backpack::base.welcome` and/or `backpack::base.use_sidebar` ❌ ❌ ❌ ❌  

What makes me a little relieved is the fact that we show this only on local, but I think we can do a little bit more to mitigate the failed scenario I referenced earlier. 

Can we check if `backpack::base.welcome` and `backpack::base.use_sidebar` matches the backpack default texts ? 
- if dashboard is published they will not get this update ✔️ 
- if it's not published and the `base.welcome` message is still the default one, I think they are not using the Backpack provided dashboard, as the default message have no relevance/meaning for a non developer user. So the previous check for the messages would mitigate our concerns.

I think with these 2 checks we are reducing by 99% the chance of showing the widget in unwanted scenarios.

What do you think ? Is it overkill with no benefits ? 

I would be more confident to say merge it, if we have some contention like that in place. 

Let me know!

Cheers

**Comment by tabacitu** [comment]
Seems like we have the same reservations here Pedro. I'm also not 100% comfortable with merging this - exactly because of the scenario we leave out, that you mentioned. But I also think what you proposed is overkill so... yeah... merged it!

the-lord-of-the-rings-mueller-report-gif-by-reactionseditor

**Comment by tabacitu** [comment]
CleanShot 2022-08-19 at 11 45 38@2x

**Pull request merged**